test/relayclient/RelayClient.test.ts

```diff
@@ -0,0 +1,829 @@
+        const finalSwBalance = await token.balanceOf(smartWallet.address);
+        const filter = testRecipient.filters.SampleRecipientEmitted();
+        const logs = await testRecipient.queryFilter(filter);
+        const log = logs.find((x) => x.transactionHash === hash);
+
+        expect(finalWorkerBalance).to.be.equal(
+          initialWorkerBalance.add(amountToBePay)
+        );
+        expect(finalSwBalance).to.be.equal(initialSwBalance.sub(amountToBePay));
+        expect(log).to.not.be.undefined;
+        expect(to).to.be.equal(relayHub.address);
+      });
+
+      it('without tokenGas estimation(not sponsored)', async function () {
+        const amountToBePay = 100;
+
+        const updatedRelayRequest = {
+          ...envelopingRelayRequest,
+          request: {
+            ...envelopingRelayRequest.request,
+            tokenAmount: amountToBePay,
+          },
+        };
+
+        const { relayWorkerAddress } = relayServer.getChainInfo();
+        const initialWorkerBalance = await token.balanceOf(relayWorkerAddress);
+        const initialSwBalance = await token.balanceOf(smartWallet.address);
+
+        const { hash, to } = await relayClient.relayTransaction(
+          updatedRelayRequest
+        );
+
+        const finalWorkerBalance = await token.balanceOf(relayWorkerAddress);
+        const finalSwBalance = await token.balanceOf(smartWallet.address);
+
+        const filter = testRecipient.filters.SampleRecipientEmitted();
+        const logs = await testRecipient.queryFilter(filter);
+        const log = logs.find((x) => x.transactionHash === hash);
+
+        expect(finalWorkerBalance).to.be.equal(
+          initialWorkerBalance.add(amountToBePay)
+        );
+        expect(finalSwBalance).to.be.equal(initialSwBalance.sub(amountToBePay));
+        expect(log).to.not.be.undefined;
+        expect(to).to.be.equal(relayHub.address);
+      });
+
+      it('with gasLimit estimation', async function () {
+        const { to: requestTo, data } = envelopingRelayRequest.request;
+
+        const { callForwarder, gasPrice } = envelopingRelayRequest.relayData;
+
+        const gasLimit = await estimateInternalCallGas({
+          to: requestTo,
+          data,
+          from: callForwarder,
+          gasPrice: (await gasPrice) ?? 0,
+        });
+
+        const updatedRelayRequest = {
+          ...envelopingRelayRequest,
+          request: {
+            ...envelopingRelayRequest.request,
+            gas: gasLimit,
+          },
+        };
+
+        const { hash, to } = await relayClient.relayTransaction(
+          updatedRelayRequest
+        );
+
+        const filter = testRecipient.filters.SampleRecipientEmitted();
+        const logs = await testRecipient.queryFilter(filter);
+        const log = logs.find((x) => x.transactionHash === hash);
+
+        expect(log).to.not.be.undefined;
+        expect(to).to.be.equal(relayHub.address);
```

Comment:
I think we could avoid this repetition in some ways (like with utility functions or high-order function)

Reply:
updated